File: frontend/src/games/mental-rotation/core/shapes.ts
/* psygames-mental-rotation-shapes · VER 2 · 09.09.2026 */
/* LOCAL REV spatial-lab/2026-09-09.1 · psygames-codex-mac · not an app release */
/**
 * БИБЛИОТЕКА ФИГУР ИЗ КУБИКОВ — общая для всех трёх видов заданий.
 *
 * Раньше набор лежал прямо в экране, и проекция считалась бы по другой копии
 * координат: два списка кубиков рано или поздно расходятся, и «правильная»
 * проекция перестаёт быть проекцией показанной фигуры. Список ОДИН, и берут его
 * и поворот, и проекция.
 *
 * ⚠️ ПЛОСКАЯ ФИГУРА — НЕ ТО ЖЕ, ЧТО КИРАЛЬНАЯ, И ЭТО РЕШАЕТ ЗАДАНИЕ. Фигуру, у
 * которой все кубики в одной плоскости, можно ПЕРЕВЕРНУТЬ в пространстве — и её
 * зеркальная копия окажется законным поворотом. Значит зеркальный отвлекающий
 * вариант на такой фигуре невозможен (он был бы вторым верным ответом), а вид
 * сверху у неё вырождается в полоску клеток. Поэтому здесь есть и плоские, и
 * объёмные фигуры, а отбор под задание идёт по свойству, а не по названию:
 * поворот просит киральных (`isChiral`), проекция — объёмных (`isVolumetric`).
 * Проба в тестах требует от библиотеки связности и попарной различимости: две
 * фигуры, оказавшиеся поворотами друг друга, сделали бы «другую фигуру» верным
 * ответом.
 */
import type { Shape, Cube } from './types';
import {createRng,pick} from './rng';
import {allOrientations,normalizeShape,shapeKey} from './geometry';

export const SHAPE_LIBRARY: readonly Shape[] = [
  // L (4)
  [[0, 0, 0], [1, 0, 0], [2, 0, 0], [2, 1, 0]],
  // Z (4)
  [[0, 0, 0], [1, 0, 0], [1, 1, 0], [2, 1, 0]],
  // T со ступенькой в глубину (4)
  [[0, 0, 0], [1, 0, 0], [1, 1, 0], [1, 1, 1]],
  // лесенка (5)
  [[0, 0, 0], [1, 0, 0], [1, 1, 0], [2, 1, 0], [2, 2, 0]],
  // ступенька в глубину (5)
  [[0, 0, 0], [1, 0, 0], [1, 0, 1], [1, 1, 1], [2, 1, 1]],
  // длинная L (5)
  [[0, 0, 0], [1, 0, 0], [2, 0, 0], [3, 0, 0], [3, 1, 0]],
  // угол с отростком вверх (5)
  [[0, 0, 0], [0, 1, 0], [0, 2, 0], [1, 2, 0], [1, 2, 1]],
  // змейка через три плоскости (6)
  [[0, 0, 0], [1, 0, 0], [1, 1, 0], [1, 1, 1], [2, 1, 1], [2, 2, 1]],
  // ветка (6)
  [[0, 0, 0], [1, 0, 0], [2, 0, 0], [2, 1, 0], [2, 1, 1], [2, 2, 1]],
  // тройник с отростком в глубину (6): единственная в библиотеке фигура с
  // РАЗВИЛКОЙ — «крюк с полкой», стоявший тут сперва, оказался поворотом ветки
  // выше, и отвлекающий вариант «другая фигура» становился вторым верным
  [[0, 0, 0], [1, 0, 0], [2, 0, 0], [1, 1, 0], [1, 1, 1], [1, 2, 1]],
  // лестница 3D (7)
  [[0, 0, 0], [1, 0, 0], [1, 1, 0], [2, 1, 0], [2, 1, 1], [3, 1, 1], [3, 2, 1]],
  // спираль (7)
  [[0, 0, 0], [1, 0, 0], [2, 0, 0], [2, 0, 1], [2, 1, 1], [2, 2, 1], [1, 2, 1]],
  // коромысло (8)
  [[0, 0, 0], [1, 0, 0], [2, 0, 0], [2, 1, 0], [2, 2, 0], [2, 2, 1], [3, 2, 1], [3, 2, 2]],
] as const as readonly Shape[];

/** Фигуры нужного размера. Пустым набор не бывает — проба в тестах это стережёт. */
export function shapesOfSize(minCubes: number, maxCubes: number): Shape[] {
  const extra:Shape[]=[];
  for(let n=Math.max(9,minCubes);n<=Math.min(13,maxCubes);n++)extra.push(...extendedShapes(n));
  return [...SHAPE_LIBRARY,...extra]
    .filter((s) => s.length >= minCubes && s.length <= maxCubes)
    .map((s) => s.map((c) => [...c] as [number, number, number]));
}

const extendedCache=new Map<number,Shape[]>();
const ADJACENT:Cube[]=[[1,0,0],[-1,0,0],[0,1,0],[0,-1,0],[0,0,1],[0,0,-1]];
/** Grow by face adjacency, keep compact extents and reject rotational duplicates.
 * Original 4..8 library remains unchanged for existing exercises/art assets. */
function extendedShapes(count:number):Shape[] {
  const cached=extendedCache.get(count);if(cached)return cached;
  const random=createRng(`rotation-shapes-${count}`),out:Shape[]=[],seen=new Set<string>();
  for(let attempt=0;attempt<300&&out.length<12;attempt++){
    let shape=SHAPE_LIBRARY[2].map(c=>[...c] as Cube);
    while(shape.length<count){
      const occupied=new Set(shape.map(c=>c.join(','))),frontier=new Map<string,Cube>();
      for(const c of shape)for(const d of ADJACENT){
        const candidate=c.map((v,i)=>v+d[i]) as Cube;
        if(occupied.has(candidate.join(',')))continue;
        if(boundingBox([...shape,candidate]).some(span=>span>4))continue;
        frontier.set(candidate.join(','),candidate);
      }
      shape.push(pick(random,[...frontier.values()]));
    }
    shape=normalizeShape(shape);
    const canonical=allOrientations(shape).map(shapeKey).sort()[0];
    if(!seen.has(canonical)){seen.add(canonical);out.push(shape);}
  }
  if(out.length<3)throw new Error(`insufficient distinct ${count}-cube shapes`);
  extendedCache.set(count,out);return out;
}

/** Габарит фигуры по осям — по нему видно, плоская она или объёмная. */
export function boundingBox(shape: Shape): [number, number, number] {
  const span = (i: 0 | 1 | 2): number => Math.max(...shape.map((c) => c[i])) - Math.min(...shape.map((c) => c[i])) + 1;
  return [span(0), span(1), span(2)];
}

/**
 * Объёмная фигура: занимает не меньше двух клеток по КАЖДОЙ оси. У плоской
 * фигуры вид сверху (или сбоку) — полоска в одну клетку шириной, и задание на
 * проекцию превращается в «сосчитай кубики в ряд».
 */
export function isVolumetric(shape: Shape): boolean {
  return boundingBox(shape).every((s) => s >= 2);
}

File: flutter/lib/games/mental_rotation/shapes.dart
/// БИБЛИОТЕКА ФИГУР — перенос `core/shapes.ts`.
///
/// Список ОДИН на все виды заданий: поворот берёт из него киральные фигуры, проекция — объёмные.
/// Две копии координат рано или поздно разъезжаются, и «правильная» проекция перестаёт быть
/// проекцией показанной фигуры.
///
/// Фигуры на 9–13 кубиков не выписаны руками, а достраиваются ростом по граням от одной заводской
/// (тот же генератор и то же семя, что в TS, — иначе полосы размеров разойдутся).
library;

import 'geometry.dart';
import 'rng.dart';

const List<List<List<int>>> shapeLibrary = [
  // L (4)
  [
    [0, 0, 0],
    [1, 0, 0],
    [2, 0, 0],
    [2, 1, 0],
  ],
  // Z (4)
  [
    [0, 0, 0],
    [1, 0, 0],
    [1, 1, 0],
    [2, 1, 0],
  ],
  // T со ступенькой в глубину (4)
  [
    [0, 0, 0],
    [1, 0, 0],
    [1, 1, 0],
    [1, 1, 1],
  ],
  // лесенка (5)
  [
    [0, 0, 0],
    [1, 0, 0],
    [1, 1, 0],
    [2, 1, 0],
    [2, 2, 0],
  ],
  // ступенька в глубину (5)
  [
    [0, 0, 0],
    [1, 0, 0],
    [1, 0, 1],
    [1, 1, 1],
    [2, 1, 1],
  ],
  // длинная L (5)
  [
    [0, 0, 0],
    [1, 0, 0],
    [2, 0, 0],
    [3, 0, 0],
    [3, 1, 0],
  ],
  // угол с отростком вверх (5)
  [
    [0, 0, 0],
    [0, 1, 0],
    [0, 2, 0],
    [1, 2, 0],
    [1, 2, 1],
  ],
  // змейка через три плоскости (6)
  [
    [0, 0, 0],
    [1, 0, 0],
    [1, 1, 0],
    [1, 1, 1],
    [2, 1, 1],
    [2, 2, 1],
  ],
  // ветка (6)
  [
    [0, 0, 0],
    [1, 0, 0],
    [2, 0, 0],
    [2, 1, 0],
    [2, 1, 1],
    [2, 2, 1],
  ],
  // тройник с отростком в глубину (6) — единственная с развилкой
  [
    [0, 0, 0],
    [1, 0, 0],
    [2, 0, 0],
    [1, 1, 0],
    [1, 1, 1],
    [1, 2, 1],
  ],
  // лестница 3D (7)
  [
    [0, 0, 0],
    [1, 0, 0],
    [1, 1, 0],
    [2, 1, 0],
    [2, 1, 1],
    [3, 1, 1],
    [3, 2, 1],
  ],
  // спираль (7)
  [
    [0, 0, 0],
    [1, 0, 0],
    [2, 0, 0],
    [2, 0, 1],
    [2, 1, 1],
    [2, 2, 1],
    [1, 2, 1],
  ],
  // коромысло (8)
  [
    [0, 0, 0],
    [1, 0, 0],
    [2, 0, 0],
    [2, 1, 0],
    [2, 2, 0],
    [2, 2, 1],
    [3, 2, 1],
    [3, 2, 2],
  ],
];

const List<List<int>> _adjacent = [
  [1, 0, 0],
  [-1, 0, 0],
  [0, 1, 0],
  [0, -1, 0],
  [0, 0, 1],
  [0, 0, -1],
];

final Map<int, List<Shape>> _extendedCache = {};

/// Фигуры нужного размера. Пустым набор не бывает — это стережёт проба.
List<Shape> shapesOfSize(int minCubes, int maxCubes) {
  final extra = <Shape>[];
  for (var n = minCubes > 9 ? minCubes : 9; n <= (maxCubes < 13 ? maxCubes : 13); n++) {
    extra.addAll(extendedShapes(n));
  }
  final all = <Shape>[
    for (final s in shapeLibrary)
      [
        for (final c in s) [...c],
      ],
    ...extra,
  ];
  return [
    for (final s in all)
      if (s.length >= minCubes && s.length <= maxCubes)
        [
          for (final c in s) [...c],
        ],
  ];
}

/// Рост по граням: компактный габарит (не шире 4) и без повторов с точностью до поворота.
List<Shape> extendedShapes(int count) {
  final cached = _extendedCache[count];
  if (cached != null) return cached;
  final random = createRng('rotation-shapes-$count');
  final out = <Shape>[];
  final seen = <String>{};
  for (var attempt = 0; attempt < 300 && out.length < 12; attempt++) {
    var shape = <Cube>[
      for (final c in shapeLibrary[2]) [...c],
    ];
    while (shape.length < count) {
      final occupied = {for (final c in shape) '${c[0]},${c[1]},${c[2]}'};
      final frontier = <String, Cube>{};
      for (final c in shape) {
        for (final d in _adjacent) {
          final candidate = <int>[c[0] + d[0], c[1] + d[1], c[2] + d[2]];
          final key = '${candidate[0]},${candidate[1]},${candidate[2]}';
          if (occupied.contains(key)) continue;
          if (boundingBox([...shape, candidate]).any((span) => span > 4)) continue;
          frontier[key] = candidate;
        }
      }
      shape.add(pick(random, frontier.values.toList()));
    }
    shape = normalizeShape(shape);
    final canonical = (allOrientations(shape).map(shapeKey).toList()..sort()).first;
    if (!seen.contains(canonical)) {
      seen.add(canonical);
      out.add(shape);
    }
  }
  if (out.length < 3) throw StateError('insufficient distinct $count-cube shapes');
  _extendedCache[count] = out;
  return out;
}
